sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala

```diff
@@ -169,11 +179,30 @@ class StreamExecution(
    * least the given `Offset`. This method is indented for use primarily when writing tests.
    */
   def awaitOffset(source: Source, newOffset: Offset): Unit = {
-    while (!currentOffsets.contains(source) || currentOffsets(source) < newOffset) {
+    def notDone = synchronized {
+      !currentOffsets.contains(source) || currentOffsets(source) < newOffset
+    }
+
+    while (notDone) {
       logInfo(s"Waiting until $newOffset at $source")
-      synchronized { wait() }
+      awaitBatchLock.synchronized { awaitBatchLock.wait(100) }
     }
     logDebug(s"Unblocked at $newOffset for $source")
   }
+
+  /** Clears the indicator that a batch has completed.  Used for testing. */
+  override def clearBatchMarker(): Unit = {
+    batchRun = false
+  }
+
+  /**
+   * Awaits the completion of at least one streaming batch. Must be called after `clearBatchMarker`
+   * to gurantee that a new batch has been processed.
+   */
+  override def awaitNextBatch(): Unit = {
+    while (!batchRun) {
+      awaitBatchLock.synchronized { awaitBatchLock.wait(100) }
```

Comment:
From the docs `interrupts and spurious wakeups are possible, and this method should always be used in a loop`.  Do my guess is that we some times wake up spuriously.  There is then a race to check the done condition / finish the batch (which is why it would hang with very low probability (3/1000)).  So, this actually does seem like the best solution.  We don't spin a ton wastefully.  In most takes we awake immediately.  In very rare cases we sleep 100ms too long.